appserver/Doc/config db schema.txt
Title:        Schema for Samoa Config Database 
Author:       Robby Slaughter
Last Updated: 1:23 PM 10/13/2003

================================================================================================================================================

==================
1.0 OVERVIEW
==================


1.1 Purpose

This document defines the database schema and supporting queries and functions for the Samoa configuration database. The target
database is SQLite 2.8.6, which is embedded within the Samoa framework.


1.2 Origin of this Feature

This database structure replaces the use of Registry keys as the primary configuration method for Samoa.


1.3 Database Structure

The structure of the database relates five key entities in the configuration and use of Samoa: these are users, components, configValues,
statistics, and extensions. The schema recognizes any number of users with distinct permissions in their use and access to components, and, 
correspondingly, configValues for these components. Furthermore, the schema identifies a relationship between configValues and particular 
components, so that any defined component can have any number of configuration values. 

Supporting tables track the use of a variety of formatTypes for the configValues. FormatTypes indicate specific interpretations of 
data, so, for example, strings can be read as file paths or descriptive names, and validating accordingly.

A statistics table permits general system statistics, as well as those relating to a particular component, to be stored and tracked by 
the system.

An additional pair of tables permits extensions to be added to components. This is done to provide a mechanism for third-party providers to 
expose additional programmable functionality to application developers. The extensionsParameters table allows these functions to have a 
variable number of parameters, with accompanying descriptions.


================================================================================================================================================


===========================================
2.0 DOCUMENT CONVENTIONS AND DATABASE CODES
===========================================

2.1 Document Conventions

This document contains SQL code intermingled with pseudocode, as instructions to the programmer. Variables specified using an __underline__ are 
intended as input parameters or values stored by the programmer between SQL calls. All SQL commands are written in UPPERCASE, with table names
specified using "camel case".

2.2 Database Codes

Several of the database fields are integer numbers which represent specific meanings. These are listed as follows:

  componentType: Integer value indicating the type of component
     0 / NULL = Core component
     1        = Provider
     2        = Application
     3        = Media Server
     4        = Native Action
     
  accessLevel: Integer field indicating an overall permission level for users

     0 = System admistrator. Complete power over the system
     1 = Power user. Can do anything except modify level 0 or other
         level 1 accounts created by the System Administrator. Can create 
         level 1, 2 or 3 subaccounts with no more power than itself.
     2 = Regular Account. Created by level 0 or 1. Only has the power to modify / view 
         config values and perform actions as specified by the creator. Can create 
         subaccounts with no more power than itself, of access level 2 or 3.
     3 = Limited Account. Created by 0, 1 or 2. Can only view data.

  accessControl: This is an integer indicating the type of access control:

     NULL = Access is undefined
     0    = All Access is prohibited
     1    = User may view the data or status, but not edit/change it
     2    = User may view and edit this data/or view/modify status

     * Note: If no record is defined, access is presumed to be level 0.


  status: An integer value for provider/application/mediaserver status
     NULL = Status is unknown
     0    = Not currently installed
     1    = Installed, but disabled
     2    = Installed and enabled
     3    = Unresponsive/offline

2.3 Access Permission Assumption

The interepration of this database is that all access permission is PERMITTED unless otherwise stated. That is, if there are no records in the
usersProvidersAccessControl table for a given user, then they are assumed to have total access to all providers. The same is the case for 
application control.

,m================================================================================================================================================

=====================
3.0 DATABASE CREATION
=====================

3.1 Database Script

The creation of the empty database can be accomplished with the following SQL script:



-- 
-- TABLE: components 
--

CREATE TABLE components(
    componentsID     INTEGER        PRIMARY KEY,
    name             CHAR(50),
    componentType    INTEGER,
    description      VARCHAR(10),
    status           INTEGER,
    version          CHAR(10),
    path             CHAR(100)
) 
;



-- 
-- TABLE: configValues 
--

CREATE TABLE configValues(
    configValuesID     INTEGER      PRIMARY KEY,
    formatTypesID      INTEGER,
    name               CHAR(255),
    componentsID       INTEGER,
    value              CHAR(255),
    description        CHAR(10),
    minValue           INTEGER,
    maxValue           INTEGER,
    diagnostic         BIT,
    requiresRestart    BIT,
    readOnly           BIT
) 
;



-- 
-- TABLE: extensions 
--

CREATE TABLE extensions(
    extensionsID    INTEGER      PRIMARY KEY,
    componentsID    INTEGER,
    name            CHAR(100),
    description     CHAR(500)
) 
;



-- 
-- TABLE: extensionsParameters 
--

CREATE TABLE extensionsParameters(
    extensionsParametersID    INTEGER      PRIMARY KEY,
    extensionsID              INTEGER,
    name                      CHAR(100),
    description               CHAR(100)
) 
;



-- 
-- TABLE: formatTypes 
--

CREATE TABLE formatTypes(
    formatTypesID    INTEGER ,
    name             CHAR(50),
    description      CHAR(50)
) 
;



-- 
-- TABLE: statistics 
--

CREATE TABLE statistics(
    statisticsID    INTEGER     PRIMARY KEY,
    componentsID    INTEGER ,
    name            CHAR(50),
    value           INTEGER
) 
;



-- 
-- TABLE: users 
--

CREATE TABLE users(
    usersID             INTEGER     PRIMARY KEY,
    userName            CHAR(20),
    password            CHAR(20),
    accessLevel         INTEGER,
    createdByUsersID    INTEGER,
    dateCreated         DATETIME,
    dateLastUpdated     DATETIME
) 
;



-- 
-- TABLE: usersComponentsAccessControl 
--

CREATE TABLE usersComponentsAccessControl(
    usersComponentsAccessControlID    INTEGER    PRIMARY KEY,
    usersID                           INTEGER,
    componentsID                      INTEGER,
    accessControl                     INTEGER
) 
;





3.2 Installation Process

The database needs to be created as part of the Samoa installation, or at least, as a step during the first runtime.
After the database is created, some basic records need to be added.

Format Types need to be inserted so that integers and strings can be handled and validated properly:


    INSERT INTO formatTypes(formatTypesID,name,description)
      VALUES (NULL,"String","Text format");

    INSERT INTO formatTypes(formatTypesID,name,description)
      VALUES (NULL,"Integer","Whole number");

    INSERT INTO formatTypes(formatTypesID,name,description)
      VALUES (NULL,"Directory","UNC path to a particular folder");

    INSERT INTO formatTypes(formatTypesID,name,description)
      VALUES (NULL,"Filename","UNC path to a particular file");

    INSERT INTO formatTypes(formatTypesID,name,description)
     VALUES (NULL,"Boolean","True/False value");

A master administrator user must be created. The password in this user should be set as part of the installation process 
so that all installs do not have a default password.

   INSERT INTO users(usersID,username, password, accessLevel, createdByUsersID)
      VALUES (NULL,"Administrator","ilovemetreos",0,NULL, NULL, NULL)


3.3 Initial Data Population

The installation of providers and default applications may require some data population. This should be done with 
interal calls that implement the SQL commands described in the next section

================================================================================================================================================


=========================
4.0 MANAGING THE DATABASE
=========================


4.1 Code Accessing the Database


Since the database is wholly embedded into Samoa, all management of data must occur through internal structures. As of 1.0, these APIs are
defined in the Samoa project Interfaces in the file IManagement.cs. Implementation of this database should only require changes to the
file referenced by this interface, which is RemotingInterface.cs in the OAM project.

4.2 User Management Functions

The database presents a variety of user management functions, not all of which need to be implemented in the initial beta release of
Samoa. These include 

4.2.1 Create A New User


Values needed:
  __Username__        - String for the username. Should be all lowercase, alphanumeric

  __Password__        - String for the password. 6-20 characters, alphanumeric + symbols

  __AccessLevel__     
  __CreatedByUsersID__ - The UsersID of the account creating this one. This is so 
                        the administrator can know where accounts came from.

Step 1: Retrieve the AccessLevel of the current user (if not already cached)
  
 SELECT AccessLevel FROM users WHERE UsersID = __CreatedByUsersID__;

Step 2: Only if the record returned by the previous step is less than the __AccessLevel__
requested, and that record is less than or equal to 2, create the new user:

 INSERT INTO users(usersID,username, password, accessLevel, createdByUsersID)
    VALUES (NULL, __Username__ , __Password__ , __AccessLevel__, __CreatedByUsersID__,
            now(), now())

Step 3: (OPTIONAL). If you want to copy all permissions from the current user onto the subuser, 
than you'll need to do this step:

    INSERT INTO usersConfigValuesAccessControl SELECT NULL,last_insert_rowid(),configValuesID,
     accessControl,now(),now();


4.2.1 Delete an Existing User


Values needed:

  __usersID__         - The usersID for the user to be deleted


Note that you cannot delete the Administrator user.

Remove all access control to applications

  DELETE FROM usersApplicationsAccessControl WHERE usersID = __usersID__;

Remove all access control to components

  DELETE FROM usersComponentsAccessControl WHERE usersID = __usersID__;

Find the parent user for this user:

  SELECT createdByUsersID FROM users WHERE usersID = __usersID__;

Delete the this user

  DELETE FROM users WHERE usersID = __usersID__;

Update all children of the (now deleted) user to refer to it's parent, repairing the chain:

  UPDATE users SET createdByUsersID = __createdByUsersID__ WHERE createdByUsersID = __usersID__;
     
          
4.3 Component Management


These sets of SQL commands enable Samoa to manage components


4.3.1 Adding a Provider


Values needed:

  __name__         - The name of the new component
  __componentType  - The type code of the component
  __description__  - A brief description of the purpose or nature of the component
  __status__ 	   - A status code for the component
  __version__      - The version number for this component
  __path__	   - A serialized System.URI object indicating the location of the component
 
  

It's wise to check if an existing component of the same name is already in the database:

   SELECT name FROM component WHERE name = __name__;

If so, then exit with an appropriate error messsage. Otherwise, insert the new provider:

   INSERT INTO providers(name,componentType,description,status,version,path) 
     VALUES (__name__,__description__,__status__,__version__,__path__);


4.3.2 Removing a Component

Values needed:
  __componetnsID__   - The id of the component to delete


(Note: SQLite has limited support for cascading delete, so we must manually adjust each table)

Step 1: Delete any configValues that are unique to this component

   DELETE FROM configValues WHERE componentsID = __componentID__;

Step 3: Delete the component itself

   DELETE FROM components WHERE componentsID = __componentsID__;
  
 
4.4 Modifying config values for Components



4.4.1 Create or Modify Config Values for a component

Values needed:

   __componentID__  - The id of the component to modify
   __name__            - The name of the configuration setting
   __value__           - The actual value, coded as a string
   __description__     - A description of the meaning of this config setting
   __minvalue__        - For numeric fields, a minimum permissable value. Otherwise, NULL
   __maxValue__        - For numeric fields, a maximum permissable value. Otherwise, NULL
   __diagnostic__      - True if this config setting is for diagnostic purposes only
   __requiresRestart__ - True if changing this setting requires a restart of Samoa 
   __readOnly__        - True if this setting CANNOT be changed.


  INSERT OR REPLACE INTO 
     configValues(componentid,name,value,description,minValue,
       maxValue,diagnostic,requiresRestart,readOnly)
       VALUES
     (__componentsID__,__name__,__value__,__description__,__minValue__,
        __maxValue__,__diagnostic__,__requiresRestart__,__readOnly__)

4.5 Changing user permissions for Components

User permissions are assumed to be ENABLED for all components, unless otherwise specified. 
(See section 2.3). To give user access to an application or a provider, change the database 
using one of the following functions.


4.5.1 Change User Permissions for a Component

Values needed:
   __usersID__           The user you want to grant/deny access
   __componentID__       The component you want to give access to
   __accessControl__     The access control 


  Note: We cannot do a INSERT OR REPLACE since no incoming fields is individually
  unique. So we, have to break it out into two queries:

  Step 1:
    SELECT usersComponentsAccessControlID FROM usersComponentsAccessControl
      WHERE usersID = __usersID__ AND componentID = __componentsID__;
   
  Step 2: 
    If the result of the previous select was no records, then
      INSERT INTO
         usersComponentsAccessControl(usersID,componentsID,accessControl)
           VALUES
         (__usersID__,__componentsID__,__accessControl__);
      
    -otherwise-      
      UPDATE usersComponentsAccessControl
         SET accessControl = __accessControl__
      WHERE
         usersComponentsAccessControlID = __usersComponentsAccessControlID__;



   
================================================================================================================================================


===========================================
5.0 APPENDIX
===========================================

5.1 Datamodel Diagram

A visual representation of the datamodel is stored with this file, using a the same name with the extension GIF.

5.2 ER Studio File

A ER Studio 6.0 File is stored with this file, using the same name with the extension DM1






Contrib/PWLib/include/ptlib/args.h
/*
 * args.h
 *
 * Program Argument Parsing class
 *
 * Portable Windows Library
 *
 * Copyright (c) 1993-1998 Equivalence Pty. Ltd.
 *
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
 * the License for the specific language governing rights and limitations
 * under the License.
 *
 * The Original Code is Portable Windows Library.
 *
 * The Initial Developer of the Original Code is Equivalence Pty. Ltd.
 *
 * Portions are Copyright (C) 1993 Free Software Foundation, Inc.
 * All Rights Reserved.
 *
 * Contributor(s): ______________________________________.
 *
 * $Log: args.h,v $
 * Revision 1.25  2005/01/26 05:37:43  csoutheren
 * Added ability to remove config file support
 *
 * Revision 1.24  2003/09/17 01:18:01  csoutheren
 * Removed recursive include file system and removed all references
 * to deprecated coooperative threading support
 *
 * Revision 1.23  2003/03/27 07:27:07  robertj
 * Added function to get a bunch of arguments as a string array.
 *
 * Revision 1.22  2002/09/16 01:08:59  robertj
 * Added #define so can select if #pragma interface/implementation is used on
 *   platform basis (eg MacOS) rather than compiler, thanks Robert Monaghan.
 *
 * Revision 1.21  2001/12/15 04:49:06  robertj
 * Added stream I/O functions for argument list.
 *
 * Revision 1.20  2000/05/25 11:05:31  robertj
 * Added PConfigArgs class so can save program arguments to config files.
 *
 * Revision 1.19  1999/03/09 02:59:49  robertj
 * Changed comments to doc++ compatible documentation.
 *
 * Revision 1.18  1999/02/16 08:07:10  robertj
 * MSVC 6.0 compatibility changes.
 *
 * Revision 1.17  1998/11/01 04:56:51  robertj
 * Added BOOl return value to Parse() to indicate there are parameters available.
 *
 * Revision 1.16  1998/10/30 11:22:14  robertj
 * Added constructors that take strings as well as const char *'s.
 *
 * Revision 1.15  1998/10/30 05:24:29  robertj
 * Added return value to << and >> operators for shifting arguments.
 *
 * Revision 1.14  1998/10/29 05:35:14  robertj
 * Fixed porblem with GetCount() == 0 if do not call Parse() function.
 *
 * Revision 1.13  1998/10/28 03:26:41  robertj
 * Added multi character arguments (-abc style) and options precede parameters mode.
 *
 * Revision 1.12  1998/10/28 00:59:46  robertj
 * New improved argument parsing.
 *
 * Revision 1.11  1998/09/23 06:20:14  robertj
 * Added open source copyright license.
 *
 * Revision 1.10  1995/12/10 11:26:38  robertj
 * Fixed signed/unsigned bug in shift count.
 *
 * Revision 1.9  1995/06/17 11:12:17  robertj
 * Documentation update.
 *
 * Revision 1.8  1995/03/14 12:40:58  robertj
 * Updated documentation to use HTML codes.
 *
 * Revision 1.7  1994/12/05  11:15:13  robertj
 * Documentation.
 *
 * Revision 1.6  1994/11/26  03:44:19  robertj
 * Documentation.
 *
 * Revision 1.6  1994/11/24  11:48:26  robertj
 * Documentation.
 *
 * Revision 1.5  1994/08/23  11:32:52  robertj
 * Oops
 *
 * Revision 1.4  1994/08/22  00:46:48  robertj
 * Added pragma fro GNU C++ compiler.
 *
 * Revision 1.3  1994/07/27  05:58:07  robertj
 * Synchronisation.
 *
 * Revision 1.2  1994/07/17  10:46:06  robertj
 * Changed to use container classes to plug memory leak.
 *
 * Revision 1.1  1994/04/01  14:08:52  robertj
 * Initial revision
 *
 */

#ifndef _PARGLIST
#define _PARGLIST

#ifdef P_USE_PRAGMA
#pragma interface
#endif

/** This class allows the parsing of a set of program arguments. This translates
   the standard argc/argv style variables passed into the main() function into
   a set of options (preceded by a '-' character) and parameters.
*/
class PArgList : public PObject
{
  PCLASSINFO(PArgList, PObject);

  public:
  /**@name Construction */
  //@{
    /** Create an argument list.
        An argument list is created given the standard arguments and a
       specification for options. The program arguments are parsed from this
       into options and parameters.

       The specification string consists of case significant letters for each
       option. If the letter is followed by the ':' character then the option
       has an associated string. This string must be in the argument or in the
       next argument.
     */
    PArgList(
      const char * theArgPtr = NULL,   /// A string constituting the arguments
      const char * argumentSpecPtr = NULL,
      /** The specification C string for argument options. See description for
         details.
       */
      BOOL optionsBeforeParams = TRUE  /// Parse options only before parameters
    );
    /** Create an argument list. */
    PArgList(
      const PString & theArgStr,   /// A string constituting the arguments
      const char * argumentSpecPtr = NULL,
      /** The specification C string for argument options. See description for
         details.
       */
      BOOL optionsBeforeParams = TRUE  /// Parse options only before parameters
    );
    /** Create an argument list. */
    PArgList(
      const PString & theArgStr,   /// A string constituting the arguments
      const PString & argumentSpecStr,
      /** The specification string for argument options. See description for
         details.
       */
      BOOL optionsBeforeParams = TRUE  /// Parse options only before parameters
    );
    /** Create an argument list. */
    PArgList(
      int theArgc,     /// Count of argument strings in theArgv
      char ** theArgv, /// An array of strings constituting the arguments
      const char * argumentSpecPtr = NULL,
      /** The specification C string for argument options. See description for
         details.
       */
      BOOL optionsBeforeParams = TRUE  /// Parse options only before parameters
    );
    /** Create an argument list. */
    PArgList(
      int theArgc,     /// Count of argument strings in theArgv
      char ** theArgv, /// An array of strings constituting the arguments
      const PString & argumentSpecStr,
      /** The specification string for argument options. See description for
         details.
       */
      BOOL optionsBeforeParams = TRUE  /// Parse options only before parameters
    );
  //@}

  /**@name Overrides from class PObject */
  //@{
    /**Output the string to the specified stream.
     */
    virtual void PrintOn(
      ostream & strm  /// I/O stream to output to.
    ) const;

    /**Input the string from the specified stream. This will read all
       characters until a end of line is reached, then parsing the arguments.
     */
    virtual void ReadFrom(
      istream & strm  /// I/O stream to input from.
    );
  //@}

  /**@name Setting & Parsing */
  //@{
    /** Set the internal copy of the program arguments.
    */
    void SetArgs(
      const PString & theArgStr /// A string constituting the arguments
    );
    /** Set the internal copy of the program arguments. */
    void SetArgs(
      int theArgc,     /// Count of argument strings in theArgv
      char ** theArgv  /// An array of strings constituting the arguments
    );
    /** Set the internal copy of the program arguments. */
    void SetArgs(
      const PStringArray & theArgs /// A string array constituting the arguments
    );

    /** Parse the arguments.
       Parse the standard C program arguments into an argument of options and
       parameters. Consecutive calls with #optionsBeforeParams# set
       to TRUE will parse out different options and parameters. If SetArgs()
       function is called then the Parse() function will restart from the
       beginning of the argument list.

       The specification string consists of case significant letters for each
       option. If the letter is followed by a '-' character then a long name
       version of the option is present. This is terminated either by a '.' or
       a ':' character. If the single letter or long name is followed by the
       ':' character then the option has may have an associated string. This
       string must be within the argument or in the next argument. If a single
       letter option is followed by a ';' character, then the option may have
       an associated string but this MUST follow the letter immediately, if
       it is present at all.

       For example, "ab:c" allows for "-a -b arg -barg -c" and
       "a-an-arg.b-option:c;" allows for "-a --an-arg --option arg -c -copt".

       @return TRUE if there is at least one parameter after parsing.
     */
    virtual BOOL Parse(
      const char * theArgumentSpec,
      /** The specification string for argument options. See description for
         details.
       */
      BOOL optionsBeforeParams = TRUE  /// Parse options only before parameters
    );
    /** Parse the arguments. */
    virtual BOOL Parse(
      const PString & theArgumentStr,
      /** The specification string for argument options. See description for
         details.
       */
      BOOL optionsBeforeParams = TRUE  /// Parse options only before parameters
    );
  //@}

  /**@name Getting parsed arguments */
  //@{
    /** Get the count of the number of times the option was specified on the
       command line.

       @return option repeat count.
     */
    virtual PINDEX GetOptionCount(
      char optionChar  /// Character letter code for the option
    ) const;
    /** Get the count of option */
    virtual PINDEX GetOptionCount(
      const char * optionStr /// String code for the option
    ) const;
    /** Get the count of option */
    virtual PINDEX GetOptionCount(
      const PString & optionName /// String code for the option
    ) const;

    /** Get if option present.
      Determines whether the option was specified on the command line.

       @return TRUE if the option was present.
     */
    BOOL HasOption(
      char optionChar  /// Character letter code for the option
    ) const;
    /** Get if option present. */
    BOOL HasOption(
      const char * optionStr /// String letter code for the option
    ) const;
    /** Get if option present. */
    BOOL HasOption(
      const PString & optionName /// String code for the option
    ) const;

    /** Get option string.
       Gets the string associated with an option e.g. -ofile or -o file
       would return the string "file". An option may have an associated string
       if it had a ':' character folowing it in the specification string passed
       to the Parse() function.

       @return the options associated string.
     */
    virtual PString GetOptionString(
      char optionChar,          /// Character letter code for the option
      const char * dflt = NULL  /// Default value of the option string
    ) const;
    /** Get option string. */
    virtual PString GetOptionString(
      const char * optionStr,   /// String letter code for the option
      const char * dflt = NULL  /// Default value of the option string
    ) const;
    /** Get option string. */
    virtual PString GetOptionString(
      const PString & optionName, /// String code for the option
      const char * dflt = NULL    /// Default value of the option string
    ) const;

    /** Get the argument count.
       Get the number of parameters that may be obtained via the
       #GetParameter()# function. Note that this does not include options
       and option strings.

       @return count of parameters.
     */
    PINDEX GetCount() const;

    /** Get the parameters that were parsed in the argument list.

       @return array of parameter strings at the specified index range.
     */
    PStringArray GetParameters(
      PINDEX first = 0,
      PINDEX last = P_MAX_INDEX
    ) const;

    /** Get the parameter that was parsed in the argument list.

       @return parameter string at the specified index.
     */
    PString GetParameter(
      PINDEX num   /// Number of the parameter to retrieve.
    ) const;

    /** Get the parameter that was parsed in the argument list. The argument
       list object can thus be treated as an "array" of parameters.

       @return parameter string at the specified index.
     */
    PString operator[](
      PINDEX num   /// Number of the parameter to retrieve.
    ) const;

    /** Shift the parameters by the specified amount. This allows the parameters
       to be parsed at the same position in the argument list "array".
     */
    void Shift(
      int sh /// Number of parameters to shift forward through list
    );

    /** Shift the parameters by the specified amount. This allows the parameters
       to be parsed at the same position in the argument list "array".
     */
    PArgList & operator<<(
      int sh /// Number of parameters to shift forward through list
    );

    /** Shift the parameters by the specified amount. This allows the parameters
       to be parsed at the same position in the argument list "array".
     */
    PArgList & operator>>(
      int sh /// Number of parameters to shift backward through list
    );
  //@}

  /**@name Errors */
  //@{
    /** This function is called when access to illegal parameter index is made
       in the GetParameter function. The default behaviour is to output a
       message to the standard #PError# stream.
     */
    virtual void IllegalArgumentIndex(
      PINDEX idx /// Number of the parameter that was accessed.
    ) const;

    /** This function is called when an unknown option was specified on the
       command line. The default behaviour is to output a message to the
       standard #PError# stream.
     */
    virtual void UnknownOption(
      const PString & option   /// Option that was illegally placed on command line.
    ) const;

    /** This function is called when an option that requires an associated
       string was specified on the command line but no associated string was
       provided. The default behaviour is to output a message to the standard
       #PError# stream.
     */
    virtual void MissingArgument(
      const PString & option  /// Option for which the associated string was missing.
    ) const;
  //@}

  protected:
    /// The original program arguments.
    PStringArray argumentArray;
    /// The specification letters for options
    PString      optionLetters;
    /// The specification strings for options
    PStringArray optionNames;
    /// The count of the number of times an option appeared in the command line.
    PIntArray    optionCount;
    /// The array of associated strings to options.
    PStringArray optionString;
    /// The index of each .
    PIntArray    parameterIndex;
    /// Shift count for the parameters in the argument list.
    int          shift;

  private:
    BOOL ParseOption(PINDEX idx, PINDEX offset, PINDEX & arg, const PIntArray & canHaveOptionString);
    PINDEX GetOptionCountByIndex(PINDEX idx) const;
    PString GetOptionStringByIndex(PINDEX idx, const char * dflt) const;
};


#ifdef P_CONFIG_FILE

/**This class parse command line arguments with the ability to override them
   from a PConfig file/registry.
  */
class PConfigArgs : public PArgList
{
    PCLASSINFO(PConfigArgs, PArgList);
  public:
  /**@name Construction */
  //@{
    PConfigArgs(
      const PArgList & args   /// Raw argument list.
    );
  //@}

  /**@name Overrides from class PArgList */
  //@{
    /** Get the count of the number of times the option was specified on the
       command line.

       @return option repeat count.
     */
    virtual PINDEX GetOptionCount(
      char optionChar  /// Character letter code for the option
    ) const;
    /** Get the count of option */
    virtual PINDEX GetOptionCount(
      const char * optionStr /// String code for the option
    ) const;
    /** Get the count of option */
    virtual PINDEX GetOptionCount(
      const PString & optionName /// String code for the option
    ) const;

    /** Get option string.
       Gets the string associated with an option e.g. -ofile or -o file
       would return the string "file". An option may have an associated string
       if it had a ':' character folowing it in the specification string passed
       to the Parse() function.

       @return the options associated string.
     */
    virtual PString GetOptionString(
      char optionChar,          /// Character letter code for the option
      const char * dflt = NULL  /// Default value of the option string
    ) const;
    /** Get option string. */
    virtual PString GetOptionString(
      const char * optionStr,   /// String letter code for the option
      const char * dflt = NULL  /// Default value of the option string
    ) const;
    /** Get option string. */
    virtual PString GetOptionString(
      const PString & optionName, /// String code for the option
      const char * dflt = NULL    /// Default value of the option string
    ) const;
  //@}

  /**@name Overrides from class PArgList */
  //@{
    /**Save the current options to the PConfig.
       This function will check to see if the option name is present and if
       so, save to the PConfig all of the arguments present in the currently
       parsed list. Note that the optionName for saving is not saved to the
       PConfig itself as this would cause the data to be saved always!
      */
    void Save(
      const PString & optionName   /// Option name for saving.
    );

    /**Set the PConfig section name for options.
      */
    void SetSectionName(
      const PString & section /// New section name
    ) { sectionName = section; }

    /**Get the PConfig section name for options.
      */
    const PString & GetSectionName() const { return sectionName; }

    /**Set the prefix for option negation.
       The default is "no-".
      */
    void SetNegationPrefix(
      const PString & prefix /// New prefix string
    ) { negationPrefix = prefix; }

    /**Get the prefix for option negation.
       The default is "no-".
      */
    const PString & GetNegationPrefix() const { return negationPrefix; }
  //@}


  protected:
    PString CharToString(char ch) const;
    PConfig config;
    PString sectionName;
    PString negationPrefix;
};

#endif // P_CONFIG_FILE

#endif

// End Of File ///////////////////////////////////////////////////////////////
